main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -99,6 +99,47 @@ object Resolver extends Phase[NamedAst.Root, ResolvedAst.Root] {
     }
   }
 
+  /**
+    * Checks that the superclasses form a DAG (no cycles).
+    */
+  // TODO improve this algo
+  private def checkSuperclassDag(classes: Map[Symbol.ClassSym, ResolvedAst.Class]): Validation[Unit, ResolutionError] = {
+
```

Comment:
I am fine with localized use of mutability, e.g. to write a simple fixpoint/cycle finding algorithm.

Reply:
In fact, I think an implementation that uses: (1) local mutable queue and (2) a mutable set of visited vertices will be much clearer.

Reply:
For the path, you can maintain a mutable map of predecessors.

Reply:
Of course, if this was Flix, we would just use Datalog constraints :)

Reply:
I refactored, but didn't use quite as much mutability as you suggested. I moved the previously implicit memoization into a mutable set of acyclic classes, and it looks much cleaner. Let me know if further change is needed.